### PR TITLE
Support for timestamps and steps in Series constructor

### PR DESCRIPTION
**Pull request by AleksanderWWW**
## Before submitting checklist

- [x] Did you **update the CHANGELOG**? (not for test updates, internal changes/refactors or CI/CD setup)
- [x] Did you **asked the docs owner** to review all the updated user-facing interfaces?

### CODE REVIEW

File: src/neptune/attributes/series/series.py [hunk 1]
```diff
@@ -116,30 +108,26 @@ def log(
...
 
-        steps = None if step is None else [step]
-        timestamps = None if timestamp is None else [timestamp] * len(value)
+        if is_stringify_value(value):
+            value = self._handle_stringified_value(value)
 
-        ops = self._get_log_operations_from_value(value, steps=steps, timestamps=timestamps)
+        if is_collection(value):
+            if step is not None and len(value) > 1:
+                raise ValueError("Collection of values are not supported for explicitly defined 'step'.")
```
Comment (normandy7) [inline src/neptune/attributes/series/series.py:121]: Verb to agree with "collection"

```suggestion
                raise ValueError("Collection of values is not supported for explicitly defined 'step'.")
```

**Review by normandy7 — COMMENTED** [review]
Thanks for tagging me!

So you can do

```
s = SomeSeries(value=..., step=..., timestamp=...)
```

or

```
s = SomeSeries(values=..., steps=..., timestamps=...)
```

and everything will work fine, as long as you pass single values to the former and equal-length collections to the latter? With only `value(s)` being required

**Comment by AleksanderWWW** [comment]
> Thanks for tagging me!
> 
> So you can do
> 
> ```
> s = SomeSeries(value=..., step=..., timestamp=...)
> ```
> 
> or
> 
> ```
> s = SomeSeries(values=..., steps=..., timestamps=...)
> ```
> 
> and everything will work fine, as long as you pass single values to the former and equal-length collections to the latter? With only `value(s)` being required

Hey, no, only `steps` and `timestamps` are viable keyword arguments in a Series constructor

**Comment by normandy7** [comment]
> Hey, no, only `steps` and `timestamps` are viable keyword arguments in a Series constructor

Got it; was lacking the context for this change. Seems it's just an internal fix/enabler for now. Thanks

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 ### Changes
 - Neptune objects and universal methods covered with docstrings ([#1309](https://github.com/neptune-ai/neptune-client/pull/1309))
+- Series objects accept `timestamps` and `steps` in their constructors ([#1318](https://github.com/neptune-ai/neptune-client/pull/1318))
```
Comment (Raalsky) [inline CHANGELOG.md:5]: Isin't it a feature?

File: src/neptune/attributes/series/file_series.py [hunk 3]
```diff
@@ -72,6 +72,9 @@ def _data_to_value(self, values: Iterable, **kwargs) -> Val:
     def _is_value_type(self, value) -> bool:
         return isinstance(value, FileSeriesVal)
 
+    def _handle_stringified_value(self, value):
```
Comment (Raalsky) [inline src/neptune/attributes/series/file_series.py:75]: As this method will be common for all serieses except String series, how about moving it to the abstract `Series` class?

File: src/neptune/internal/types/stringify_value.py [hunk 4]
```diff
@@ -39,6 +39,12 @@ class StringifyValue:
     def __init__(self, value: Any):
         self.__value = value
 
+    def __len__(self) -> int:
```
Comment (Raalsky) [inline src/neptune/internal/types/stringify_value.py:42]: I don't like it :P. `StringifyValue` may be a single value as well and it doesn't make any sense. Think of it as some kind of a "serialization", if you need to access the underlying value -> unpack it.

File: src/neptune/internal/types/stringify_value.py [hunk 5]
```diff
@@ -39,6 +39,9 @@ class StringifyValue:
     def __init__(self, value: Any):
         self.__value = value
 
+    def __getitem__(self, item):
```
Comment (Raalsky) [inline src/neptune/internal/types/stringify_value.py:42]: The same applies here, you cannot assume that it's an array all the time :wink:.

Comment (AleksanderWWW) [reply src/neptune/internal/types/stringify_value.py:42]: Ok, but shall I leave them in `tests/unit/neptune/new/test_stringify_unsupported.py`? There we have a type hint of `str` in the constructor, so those methods will make sense for an underlying value of type string.

File: src/neptune/attributes/series/series.py [hunk 6]
```diff
@@ -50,23 +47,12 @@ def __init_subclass__(cls, max_batch_size: int, operation_cls: type(LogOperation
         cls.max_batch_size = max_batch_size
         cls.operation_cls = operation_cls
 
-    def clear(self, *, wait: bool = False) -> None:
+    def clear(self, wait: bool = False) -> None:
```
Comment (Raalsky) [inline src/neptune/attributes/series/series.py:50]: Why do we need to remove the `*` here?

File: tests/unit/neptune/new/attributes/series/test_float_series.py [hunk 7]
```diff
@@ -51,3 +52,7 @@ def test_log(self):
             values = list(var.fetch_values()["value"].array)
             expected = list(range(0, 5000))
             self.assertEqual(len(set(expected)), len(set(values)))
+
+    def test_handle_stringified_value(self):
```
Comment (Raalsky) [inline tests/unit/neptune/new/attributes/series/test_float_series.py:56]: This looks a little bit weird as we're mocking some internal value. Isin't it already covered in stringify value related test?

File: tests/unit/neptune/new/test_stringify_unsupported.py [hunk 8]
```diff
@@ -44,6 +44,12 @@ class Obj:
     def __init__(self, name: str = "A"):
         self._name = name
 
+    def __len__(self) -> int:
```
Comment (Raalsky) [inline tests/unit/neptune/new/test_stringify_unsupported.py:47]: Use a list of objects instead in tests. `Obj` should be just a single value and really similar to dataclass (or even we can made it a dataclass, but no need to do to my knowledge)

File: src/neptune/internal/types/stringify_value.py (hunk 5 above)
Comment (Raalsky) [reply src/neptune/internal/types/stringify_value.py]: No, `Obj` should be a really simple value like `int` or `string` but not a built-in :wink:. `len` and index-based access looks like an array/list.

File: src/neptune/attributes/series/series.py (hunk 6 above)
Comment (AleksanderWWW) [reply src/neptune/attributes/series/series.py:50]: By mistake! My bad

File: src/neptune/attributes/series/string_series.py [hunk 9]
```diff
@@ -65,20 +65,31 @@ def _get_log_operations_from_value(
...
         if kwargs:
             logger.warning("Warning: unexpected arguments (%s) in StringSeries", kwargs)
 
-        return StringSeriesVal(values)
+        return StringSeriesVal(values, steps=steps, timestamps=timestamps)
 
     def _is_value_type(self, value) -> bool:
         return isinstance(value, StringSeriesVal)
 
+    def _handle_stringified_value(self, value):
```
Comment (Raalsky) [inline src/neptune/attributes/series/string_series.py:85]: Just a code style suggestion:
```python
    def _handle_stringified_value(self, value):
        if is_collection(value.value):
            return list(map(str, value.value))
        return str(value.value)
```

**Comment by Raalsky** [comment]
E2E tests: https://github.com/neptune-ai/neptune-client/actions/runs/4691318134

**Review by Raalsky — APPROVED** [review]

**Pull request merged**